kv/kvclient/kvcoord: TestMuxRangeFeedMetricsManagement failed

kv/kvclient/kvcoord.TestMuxRangeFeedMetricsManagement [failed](https://mesolite.cluster.engflow.com/invocations/default/07c49dcf-90bb-426d-a596-5315907d10f8?testReportRun=2&testReportShard=13&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZjbGllbnQva3Zjb29yZDprdmNvb3JkX3Rlc3Q=) on release-26.1 @ [56b2b811b3d49b5ac77ec186243c373187bd5f3d](https://github.com/cockroachdb/cockroach/commits/56b2b811b3d49b5ac77ec186243c373187bd5f3d):


```
=== RUN   TestMuxRangeFeedMetricsManagement
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestMuxRangeFeedMetricsManagement615682622
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:123: cluster virtualization disabled in global scope due to issue: #156327 (expected label: C-bug)
    test_server_shim.go:668: DRPC is enabled (via override by TestingGlobalDRPCOption)
    dist_sender_rangefeed_test.go:585: 
        	Error Trace:	pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go:585
        	Error:      	Not equal: 
        	            	expected: 2
        	            	actual  : 3
        	Test:       	TestMuxRangeFeedMetricsManagement
    panic.go:615: -- test log scope end --
test logs left over in: outputs.zip/logTestMuxRangeFeedMetricsManagement615682622
--- FAIL: TestMuxRangeFeedMetricsManagement (5.42s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=13</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMuxRangeFeedMetricsManagement.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-58220

--- a/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender_rangefeed_test.go
@@ -10,6 +10,7 @@ import (
 	"fmt"
 	"math/rand"
 	"reflect"
+	"sync"
 	"sync/atomic"
 	"testing"
 	"time"
@@ -510,7 +511,13 @@ func TestMuxRangeFeedMetricsManagement(t *testing.T) {
 		return skipSet.stuck.Contains(k)
 	}
 
-	frontierAdvanced := make(chan struct{}, 1)
+	var closeOnce sync.Once
+	frontierAdvanced := make(chan struct{})
+	frontierHasAdvanced := func() {
+		closeOnce.Do(func() {
+			close(frontierAdvanced)
+		})
+	}
 	ignoreValues := func(event kvcoord.RangeFeedMessage) {}
 	errCh, closeFeed := rangeFeed(ts.DistSenderI(), fooSpan, startTime, ignoreValues,
 		kvcoord.TestingWithRangeFeedMetrics(&metrics),
@@ -549,10 +556,7 @@ func TestMuxRangeFeedMetricsManagement(t *testing.T) {
 							return false, err
 						}
 						if advanced {
-							select {
-							case frontierAdvanced <- struct{}{}:
-							default:
-							}
+							defer frontierHasAdvanced()
 						}
 
 						if numCatchupBlocked.Add(1) <= numCatchupToBlock {
@@ -571,8 +575,9 @@ func TestMuxRangeFeedMetricsManagement(t *testing.T) {
 			}))
 	defer closeFeed()
 
-	// Wait for the test frontier to advance.  Once it advances,
-	// we know the rangefeed is started, all ranges are running (even if some of them are blocked).
+	// Wait for the test frontier to advance. Once it advances, we know the
+	// rangefeed is started, all ranges are running (even if some of them are
+	// blocked).
 	channelWaitWithTimeout(t, frontierAdvanced, errCh)
 
 	// At this point, we know the rangefeed for all ranges are running.
@@ -581,8 +586,24 @@ func TestMuxRangeFeedMetricsManagement(t *testing.T) {
 	// All ranges expected to be local.
 	require.EqualValues(t, numRanges, metrics.RangefeedLocalRanges.Value(), frontier.String())
 
-	// We also know that we have blocked numCatchupToBlock ranges in their catchup scan.
-	require.EqualValues(t, numCatchupToBlock, metrics.RangefeedCatchupRanges.Value())
+	// We also know that we have blocked numCatchupToBlock ranges from emitting
+	// the checkpoint that would allow them to mark their catch-up scan as done.
+	//
+	// We use succeed soon here because we could observe extra ranges in catch-up
+	// scan if our testing hook _just_ observed the checkpoint such that we were
+	// unblocked from channelWaitWithTimeout above but the metric hasn't been
+	// decremented yet.
+	//
+	// We should never observed fewer ranges in catchup, so test for that case
+	// without waiting.
+	require.GreaterOrEqual(t, metrics.RangefeedCatchupRanges.Value(), numCatchupToBlock)
+	testutils.SucceedsSoon(t, func() error {
+		if count := metrics.RangefeedCatchupRanges.Value(); count != numCatchupToBlock {
+			return errors.Errorf("expected %d catchup ranges, got %d",
+				numCatchupToBlock, count)
+		}
+		return nil
+	})
 }
 
 // TestMuxRangefeedRangeObserver ensures the kvcoord.WithRangeObserver option

